File: frontend/src/types/packing.ts
export interface ItemCreate {
  supplier_nombre?: string
  supplier_numero?: string
  foto_url?: string
  // Foto final (limpia) para los documentos de cliente/proveedor; si falta se usa foto_url.
  foto_final_url?: string | null
  // Marca de fabrica y fecha de entrega del proveedor: los llena la vendedora
  marca?: string | null
  fecha_recibo?: string | null
  item_no?: string
  descripcion_es?: string
  descripcion_en?: string
  descripcion_zh?: string
  material?: string
  uso?: string
  qty_por_ctn: number
  price_rmb: number
  gw?: number
  largo_cm?: number
  ancho_cm?: number
  alto_cm?: number
  // CBM directo de etiqueta (opcional); si no, se calcula por dimensiones.
  cbm?: number | null
  // MQT: mínima cantidad de cajas que pide el proveedor.
  moq_cajas?: number | null
  ctns: number
  orden?: number
  // Colores/variantes: el OCR ya lo lee siempre, no solo en bolsos.
  colores?: string | null
  // Campos propios de una cotización de bolsos (sesion.tipo_cotizacion === 'bolsos').
  tamano?: string | null
  empaque?: string | null
  etiqueta?: string | null
  herrajes?: string | null
  riata?: string | null
  // Mínimo que exige la TIENDA en total: puede ser solo cajas, o cajas + piezas.
  minimo_cajas_tienda?: number | null
  minimo_piezas_caja_tienda?: number | null
  // Fotos de detalle del bolso: {"interior": url, "herrajes": url, ...}
  fotos_extra?: Record<string, string> | null
  // Recortes a mano de las de arriba, hechos ya durante la carga masiva (antes
  // de agregar el producto): sin esto se perdían al pasar de resultado de OCR
  // a ítem real de la cotización.
  fotos_extra_final?: Record<string, string> | null
  // Si viene de un resultado de carga masiva: marca ese resultado como ya
  // agregado para que no se ofrezca ni se duplique si se retoma el lote.
  lote_item_id?: string
}

export interface ItemResponse extends ItemCreate {
  id: string
  sesion_id: string
  foto_url?: string
  // Referencia de catálogo que ve el cliente (la asigna el sistema)
  referencia?: string | null
  cantidad_solicitada?: number | null
  t_qty: number
  total_rmb: number
  price_usd: number
  total_usd: number
  cbm: number
  t_cbm: number
  t_gw: number
}

export interface Sesion {
  id: string
  nombre_cliente: string
  fecha: string
  // "productos" (default) o "bolsos": cambia qué datos pide el OCR.
  tipo_cotizacion: 'productos' | 'bolsos'
  tipo_cambio_usd: number
  user_id: string
  cliente_id?: string | null
  contenedor_id?: string | null
  // Marca de embarque: identifica la carga del cliente dentro del contenedor
  shipping_mark?: string | null
  enviada_cliente?: boolean
  notas_cliente?: string | null
  pedido_recibido_at?: string | null
  pedido_estado?: string | null
  pedido_confirmado_at?: string | null
  orden_compra_url?: string | null
  orden_compra_nombre?: string | null
  created_at: string
  // Etapa real del envío (ESTADOS_ENVIO backend), null si aún no tiene
  // seguimiento (todavía no llegó a "cotización enviada").
  estado_envio?: string | null
  // Con estado_envio === 'en_bodega' a secas no se distingue "bodega la está
  // revisando" de "bodega ya se la envió al cliente a aprobar": estos dos
  // campos son los que permiten esa distinción.
  cliente_aprobo_despacho_at?: string | null
  aprobacion_limite_at?: string | null
}

File: backend/app/schemas/packing.py
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict


class ItemCreate(BaseModel):
    """Datos crudos para crear un ítem del packing list"""

    supplier_nombre: Optional[str] = None
    supplier_numero: Optional[str] = None
    foto_url: Optional[str] = None
    # Foto final (limpia) para los documentos de cliente/proveedor; si falta se usa foto_url.
    foto_final_url: Optional[str] = None
    item_no: Optional[str] = None
    descripcion_es: Optional[str] = None
    descripcion_en: Optional[str] = None
    descripcion_zh: Optional[str] = None
    material: Optional[str] = None
    uso: Optional[str] = None
    qty_por_ctn: int = 1
    price_rmb: float = 0.0
    gw: float = 0.0
    largo_cm: float = 0.0
    ancho_cm: float = 0.0
    alto_cm: float = 0.0
    # CBM directo de etiqueta (opcional); si es None se calcula por dimensiones.
    cbm: Optional[float] = None
    # MQT: mínima cantidad de cajas que pide el proveedor.
    moq_cajas: Optional[int] = None
    ctns: int = 1
    orden: int = 0


class RecorteRequest(BaseModel):
    """Recuadro para recortar a mano la foto de un producto.

    [x0, y0, x1, y1] en fracciones de 0 a 1. En null se descarta el recorte y los
    documentos vuelven a usar la foto completa.
    """

    recuadro: Optional[list[float]] = None


class ItemUpdate(BaseModel):
    """Campos opcionales para actualización parcial (PATCH) de un ítem"""

    supplier_nombre: Optional[str] = None
    supplier_numero: Optional[str] = None
    foto_final_url: Optional[str] = None
    item_no: Optional[str] = None
    descripcion_es: Optional[str] = None
    descripcion_en: Optional[str] = None
    descripcion_zh: Optional[str] = None
    material: Optional[str] = None
    uso: Optional[str] = None
    qty_por_ctn: Optional[int] = None
    price_rmb: Optional[float] = None
    gw: Optional[float] = None
    largo_cm: Optional[float] = None
    ancho_cm: Optional[float] = None
    alto_cm: Optional[float] = None
    cbm: Optional[float] = None
    moq_cajas: Optional[int] = None
    ctns: Optional[int] = None
    orden: Optional[int] = None


class ItemResponse(ItemCreate):
    """Ítem completo con sus campos calculados (no almacenados en DB)"""

    id: str
    sesion_id: str
    foto_url: Optional[str] = None
    referencia: Optional[str] = None  # referencia de catálogo (documentos del cliente)
    cantidad_solicitada: Optional[int] = None  # cajas que pidió el cliente en su portal
    t_qty: int
    total_rmb: float
    price_usd: float
    total_usd: float
    cbm: float
    t_cbm: float
    t_gw: float

    model_config = ConfigDict(from_attributes=True)


class SesionCreate(BaseModel):
    """Datos para crear una sesión de cotización"""

    nombre_cliente: str
    tipo_cambio_usd: float = 6.7
    # Si se crea para un cliente del portal, su id (la cotización nace vinculada)
    cliente_id: str | None = None


class SesionResponse(BaseModel):
    """Sesión de cotización"""

    id: str
    nombre_cliente: str
    fecha: date
    tipo_cambio_usd: float
    user_id: str
    cliente_id: str | None = None
    contenedor_id: str | None = None  # embarque al que pertenece (define la TRM al facturar)
    enviada_cliente: bool = False
    notas_cliente: str | None = None  # observaciones que dejó el cliente en su pedido
    pedido_recibido_at: datetime | None = None  # cuándo el cliente envió su pedido
    pedido_estado: str | None = None  # recibido / por_confirmar / confirmado
    pedido_confirmado_at: datetime | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ReordenarItem(BaseModel):
    """Par id/orden para reordenar ítems"""

    id: str
    orden: int


class CantidadClienteLinea(BaseModel):
    """Cajas que se piden de un producto (cantidad final del cliente)"""

    item_id: str
    cantidad: int


class EnviarAConfirmarInput(BaseModel):
    """La vendedora ajusta las cantidades y las envía al cliente a confirmar"""

    items: list[CantidadClienteLinea]
